File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head == null) return null;
        
        ListNode dummy = new ListNode();
        dummy.next = head;
        
        ListNode slow = dummy;
        ListNode fast = dummy;
        
        boolean flag = false;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            
            if(slow == fast){
                flag = true;
                break;
            }    
        }

        if(flag){
            fast = dummy;
            while(fast != slow){
                slow = slow.next;
                fast = fast.next;
            }
            return fast;
        }
        
        return null;
    }
}

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(head == NULL) return NULL;
        
        ListNode *dummy = new ListNode();
        dummy->next = head;
        
        ListNode *slow = dummy;
        ListNode *fast = dummy;
        
        while(fast != NULL && fast->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
            
            if(slow == fast){
                fast = dummy;
                while(fast != slow){
                    slow = slow->next;
                    fast = fast->next;
                 }
                 return fast;
            }    
        }

        
        return NULL;
    }
};